freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UsersStateDifference.java

```diff
@@ -78,28 +81,60 @@ public static UsersStateDifference fromUmsAndIpaUsersStates(UsersState umsState,
 
         Multimap<String, String> groupMembershipToAdd = HashMultimap.create();
         umsGroupMembership.forEach((group, user) -> {
-            LOGGER.info("Evaluation group = {} and user = {}", group, user);
             if (!ipaGroupMembership.containsEntry(group, user)) {
-                LOGGER.info("adding");
+                LOGGER.debug("adding user : {} to group : {}", user, group);
                 groupMembershipToAdd.put(group, user);
             }
         });
+        LOGGER.info("groupMembershipToAdd size= {}", groupMembershipToAdd.size());
 
         Multimap<String, String> groupMembershipToRemove = HashMultimap.create();
         ipaGroupMembership.forEach((group, user) -> {
-            LOGGER.info("Evaluation group = {} and user = {}", group, user);
             if (!umsGroupMembership.containsEntry(group, user)) {
-                LOGGER.info("removing");
+                LOGGER.debug("removing user : {} to group : {}", user, group);
                 groupMembershipToRemove.put(group, user);
             }
         });
 
+        LOGGER.info("groupMembershipToRemove size= {}", groupMembershipToRemove.size());
+
+        Set<FmsUser> usersToRemove =
+            getUsersToBeRemoved(umsState.getGroupMembership().get(UserServiceConstants.CDP_USERSYNC_INTERNAL_GROUP),
+                                ipaState.getUsers());
+        LOGGER.info("usersToRemove size= {}", usersToRemove.size());
+
+        Set<FmsGroup> groupsToBeRemoved = getGroupsToBeRemoved(umsState.getGroups(), ipaState.getGroups());
         return new UsersStateDifference(
-                Set.copyOf(Sets.difference(umsState.getGroups(), ipaState.getGroups())),
-                Set.copyOf(Sets.difference(umsState.getUsers(), ipaState.getUsers())),
-                Set.copyOf(Sets.difference(ipaState.getGroups(), umsState.getGroups())),
-                Set.copyOf(Sets.difference(ipaState.getUsers(), umsState.getUsers())),
-                groupMembershipToAdd,
-                groupMembershipToRemove);
+            Set.copyOf(Sets.difference(umsState.getGroups(), ipaState.getGroups())),
+            Set.copyOf(Sets.difference(umsState.getUsers(), ipaState.getUsers())),
+            groupsToBeRemoved,
+            usersToRemove,
+            groupMembershipToAdd,
+            groupMembershipToRemove);
+    }
+
+    private static Set<FmsGroup> getGroupsToBeRemoved(Set<FmsGroup> umsGroups, Set<FmsGroup> ipaGroups) {
+        Set<FmsGroup> groupsToBeRemoved = new HashSet<>();
+
+        ipaGroups.forEach(group -> {
+            if (!umsGroups.contains(group) && !group.getName().equals("admins")) {
+                groupsToBeRemoved.add(group);
+            }
+        });
+
+        return groupsToBeRemoved;
+
+
+    }
+
+    private static Set<FmsUser> getUsersToBeRemoved(Collection<String> users, Set<FmsUser> ipaStateUsers) {
```

Comment:
I don't think this is right. I read this as:
Remove users that 1. exist in ipa, and 2. do not exist in ums.
This is because all UMS users will be in the internal group. 

I think we want logic that says:
Remove users that 1. exist in ipa, and 2. are part of the internal group in ipa, and 3. do not exist in ums

Reply:
Good Catch, I have changed this.